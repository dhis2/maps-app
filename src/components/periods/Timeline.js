import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { select } from 'd3-selection';
import { axisBottom } from 'd3-axis';
import { scaleTime } from '../../util/periods';

const styles = theme => ({
    root: {
        backgroundColor: 'rgba(255,255,255,0.9)',
        boxShadow: theme.shadows[1],
        borderRadius: theme.shape.borderRadius,
        width: 'calc(100% - 20px)',
        position: 'absolute',
        left: 10,
        bottom: 28,
        height: 48,
        zIndex: 1000,
        userSelect: 'none',
    },
    play: {
        cursor: 'pointer',
        fill: '#333',
        '&:hover': {
            fill: '#000',
        },
    },
    period: {
        fill: '#333',
        stroke: '#555',
        strokeWidth: 1,
        fillOpacity: 0.1,
        cursor: 'pointer',
        '&.selected': {
            fill: '#2b675c',
            fillOpacity: 0.9,
        },
        '&:hover': {
            '&:not(.selected)': {
                fillOpacity: 0.2,
            },
        },
    },
});

const paddingLeft = 40;
const paddingRight = 20;
const labelWidth = 80;
const delay = 1500;
const playBtn = <path d="M8 5v14l11-7z" />;
const pauseBtn = <path d="M6 19h4V5H6v14zm8-14v14h4V5h-4z" />;
const doubleTicksPeriods = ['LAST_6_BIMONTHS', 'BIMONTHS_THIS_YEAR'];

export class Timeline extends Component {
    static contextTypes = {
        map: PropTypes.object,
    };

    static propTypes = {
        periodId: PropTypes.string.isRequired,
        period: PropTypes.object.isRequired,
        periods: PropTypes.array.isRequired,
        onChange: PropTypes.func.isRequired,
        classes: PropTypes.object.isRequired,
    };

    state = {
        width: null,
        mode: 'start',
    };

    componentDidMount() {
        this.setWidth();
        this.context.map.on('resize', this.setWidth);
    }

    componentDidUpdate() {
        this.setTimeAxis();
    }

    componentWillUnmount() {
        this.context.map.off('resize', this.setWidth);
    }

    render() {
        const { classes } = this.props;
        const { mode } = this.state;

        this.setTimeScale();

        return (
            <svg className={`dhis2-map-timeline ${classes.root}`}>
                <g
                    onClick={this.onPlayPause}
                    transform="translate(7,5)"
                    className={classes.play}
                >
                    <path d="M0 0h24v24H0z" fillOpacity="0.0" />
                    {mode === 'play' ? pauseBtn : playBtn}
                </g>
                <g transform={`translate(${paddingLeft},10)`}>
                    {this.getPeriodRects()}
                </g>
                <g
                    transform={`translate(${paddingLeft},25)`}
                    ref={node => (this.node = node)}
                />
            </svg>
        );
    }

    // Returns array of period rectangles
    getPeriodRects = () => {
        const { period, periods, classes } = this.props;

        return periods.map(item => {
            const isCurrent = period.id === item.id;
            const { id, startDate, endDate } = item;
            const x = this.timeScale(startDate);
            const width = this.timeScale(endDate) - x;
            const className = `${classes.period}${
                isCurrent ? ` selected` : ''
            }`;

            return (
                <rect
                    key={id}
                    className={className}
                    x={x}
                    y="0"
                    width={width}
                    height="16"
                    onClick={() => this.onPeriodClick(item)}
                />
            );
        });
    };

    // Set time scale
    setTimeScale = () => {
        const { periods } = this.props;
        const { width } = this.state;

        if (!periods.length) {
            return;
        }

        const { startDate } = periods[0];
        const { endDate } = periods[periods.length - 1];

        // Link time domain to timeline width
        this.timeScale = scaleTime()
            .domain([startDate, endDate])
            .range([0, width]);
    };

    // Set timeline axis
    setTimeAxis = () => {
        const { periodId, periods } = this.props;
        const numPeriods =
            periods.length * (doubleTicksPeriods.includes(periodId) ? 2 : 1);
        const { width } = this.state;
        const ticks = Math.round(width / labelWidth);
        const timeAxis = axisBottom(this.timeScale);

        timeAxis.ticks(ticks < numPeriods ? ticks : numPeriods);

        select(this.node).call(timeAxis);
    };

    // Set timeline width from DOM el
    setWidth = () => {
        if (this.node) {
            // clientWith returns 0 for SVG elements in Firefox
            const box = this.node.parentNode.getBoundingClientRect();
            const width = box.right - box.left - paddingLeft - paddingRight;

            this.setState({ width });
        }
    };

    // Handler for period click
    onPeriodClick(period) {
        // Switch to period if different
        if (period.id !== this.props.period.id) {
            this.props.onChange(period);
        }

        // Stop animation if running
        this.stop();
    }

    // Handler for play/pause button
    onPlayPause = () => {
        if (this.state.mode === 'play') {
            this.stop();
        } else {
            this.play();
        }
    };

    // Play animation
    play = () => {
        const { period, periods, onChange } = this.props;
        const index = periods.findIndex(p => p.id === period.id);
        const isLastPeriod = index === periods.length - 1;

        // If new animation
        if (!this.timeout) {
            // Switch to first period if last
            if (isLastPeriod) {
                onChange(periods[0]);
            }

            this.setState({ mode: 'play' });
        } else {
            // Stop animation if last period
            if (isLastPeriod) {
                this.stop();
                return;
            }

            // Switch to next period
            onChange(periods[index + 1]);
        }

        // Call itself after delay
        this.timeout = setTimeout(this.play, delay);
    };

    // Stop animation
    stop = () => {
        this.setState({ mode: 'stop' });
        clearTimeout(this.timeout);
        delete this.timeout;
    };
}

export default withStyles(styles)(Timeline);
