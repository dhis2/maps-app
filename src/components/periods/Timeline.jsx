import cx from 'classnames'
import { axisBottom } from 'd3-axis'
import { scaleTime } from 'd3-scale'
import { select } from 'd3-selection'
import PropTypes from 'prop-types'
import React, { useState, useEffect, useRef, useCallback, useMemo } from 'react'
import {
    addPeriodsDetails,
    sortPeriodsByLevelAndStartDate,
    checkLastPeriod,
    getMinMaxDates,
} from '../../util/periods.js'
import timeTicks from '../../util/timeTicks.js'
import styles from './styles/Timeline.module.css'

const PADDING_LEFT = 40
const PADDING_RIGHT = 20
const LABEL_WIDTH = 80
const RECT_HEIGHT = 8
const RECT_OFFSET = 8
const DELAY = 1500
const MODE_PLAY = 'play'
const MODE_PAUSE = 'pause'

const Timeline = ({ period, periods, onChange, resizeCount }) => {
    const [width, setWidth] = useState(null)
    const [mode, setMode] = useState('start')
    const [currentPeriod, setCurrentPeriod] = useState(period)
    const svgRef = useRef(null)
    const timeoutRef = useRef(null)

    const TRANSPARENT_RECT = <path d="M0 0h24v24H0z" fillOpacity="0.0" />
    const PLAY_ICON = (
        <path d="M8 5v14l11-7z" className="play-icon" data-test="play-button" />
    )
    const PAUSE_ICON = (
        <path
            d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
            className="pause-icon"
            data-test="pause-button"
        />
    )

    const updateWidth = useCallback(() => {
        if (svgRef.current) {
            const box = svgRef.current.parentNode?.getBoundingClientRect()
            const newWidth = box ? box.width - PADDING_LEFT - PADDING_RIGHT : 0
            setWidth(newWidth)
        }
    }, [])

    useEffect(() => {
        window.addEventListener('resize', updateWidth)
        updateWidth()
        return () => window.removeEventListener('resize', updateWidth)
    }, [updateWidth, resizeCount])

    const { periodsWithTypeLevelAndRank, distinctLevelsCount } = useMemo(
        () => addPeriodsDetails(periods),
        [periods]
    )

    const sortedPeriods = useMemo(
        () => sortPeriodsByLevelAndStartDate(periodsWithTypeLevelAndRank),
        [periodsWithTypeLevelAndRank]
    )

    const isLastPeriod = useMemo(
        () => checkLastPeriod(currentPeriod, sortedPeriods),
        [currentPeriod, sortedPeriods]
    )

    const timeScale = useMemo(() => {
        if (!sortedPeriods.length || !width) {
            return null
        }
        return scaleTime()
            .domain(getMinMaxDates(sortedPeriods))
            .range([0, width])
    }, [sortedPeriods, width])

    const setTimeAxis = useCallback(() => {
        if (!timeScale) {
            return
        }
        const maxTicks = Math.round(width / LABEL_WIDTH)
        const ticks = timeTicks(...timeScale.domain(), maxTicks)
        const timeAxis = axisBottom(timeScale).tickValues(ticks)
        select(svgRef.current).call(timeAxis)
    }, [timeScale, width])

    useEffect(() => {
        setTimeAxis()
    }, [timeScale, setTimeAxis])

    const play = useCallback(() => {
        timeoutRef.current = setTimeout(() => {
            const currentPeriodIndex = sortedPeriods.findIndex(
                (period) => period.id === currentPeriod.id
            )
            const nextPeriodIndex = currentPeriodIndex + 1
            if (nextPeriodIndex < sortedPeriods.length) {
                setCurrentPeriod(sortedPeriods[nextPeriodIndex])
                onChange(sortedPeriods[nextPeriodIndex])
            } else {
                setMode(MODE_PAUSE)
            }
        }, DELAY)
        setMode(MODE_PLAY)
    }, [sortedPeriods, currentPeriod, onChange])

    const pause = useCallback(() => {
        clearTimeout(timeoutRef.current)
        setMode(MODE_PAUSE)
    }, [])

    useEffect(() => {
        mode === MODE_PLAY ? play() : pause()
        return () => clearTimeout(timeoutRef.current)
    }, [mode, play, pause])

    const onPlayPause = useCallback(() => {
        mode === MODE_PLAY ? pause() : play()
        if (isLastPeriod) {
            setCurrentPeriod(sortedPeriods[0])
            onChange(sortedPeriods[0])
        }
    }, [mode, isLastPeriod, play, pause, sortedPeriods, onChange])

    const onPeriodClick = (clickedPeriod) => {
        if (clickedPeriod.id !== period.id) {
            setCurrentPeriod(clickedPeriod)
            onChange(clickedPeriod)
        }
        pause()
    }

    const rectTotalHeight =
        RECT_HEIGHT + (distinctLevelsCount - 1) * RECT_OFFSET

    const getPeriodRects = (sortedPeriods, timeScale) => {
        return sortedPeriods.map((item) => {
            const isCurrent = period.id === item.id
            const x = timeScale(new Date(item.startDate))
            const width = timeScale(new Date(item.endDate)) - x

            return (
                <rect
                    key={item.id}
                    className={cx(styles.period, {
                        [styles.selected]: isCurrent,
                    })}
                    x={x}
                    y={item.levelRank * RECT_OFFSET}
                    width={width}
                    height={RECT_HEIGHT}
                    onClick={() => onPeriodClick(item)}
                    role="button"
                />
            )
        })
    }

    return (
        <svg
            className={`dhis2-map-timeline ${styles.timeline}`}
            style={{
                height: `${32 + rectTotalHeight}px`,
                bottom: '30px',
            }}
        >
            {/* Play/Pause Button */}
            <g
                onClick={onPlayPause}
                transform={`translate(7, ${rectTotalHeight / 2})`}
                className={styles.play}
                role="button"
                aria-label="Play/Pause"
            >
                {TRANSPARENT_RECT}
                {mode === MODE_PLAY ? PAUSE_ICON : PLAY_ICON}
            </g>
            {/* Period Rectangles */}
            {sortedPeriods && timeScale && (
                <g transform={`translate(${PADDING_LEFT}, 10)`}>
                    {getPeriodRects(sortedPeriods, timeScale)}
                </g>
            )}
            {/* X-Axis */}
            <g
                transform={`translate(${PADDING_LEFT}, ${
                    12 + rectTotalHeight
                })`}
                ref={svgRef}
            />
        </svg>
    )
}

Timeline.propTypes = {
    period: PropTypes.object.isRequired,
    periods: PropTypes.array.isRequired,
    onChange: PropTypes.func.isRequired,
    resizeCount: PropTypes.number,
}

export default Timeline
