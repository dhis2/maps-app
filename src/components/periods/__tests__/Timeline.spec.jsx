import { render, screen, fireEvent } from '@testing-library/react'
import React from 'react'
import Timeline from '../Timeline.jsx'

describe('Timeline', () => {
    const periodId = 'LAST_3_MONTHS'
    const period = { id: '201906', name: 'June 2019' }
    const periods = [
        {
            id: '201904',
            name: 'April 2019',
            startDate: new Date('2019-04-01T00:00:00.000'),
            endDate: new Date('2019-04-30T24:00:00.000'),
        },
        {
            id: '201905',
            name: 'May 2019',
            startDate: new Date('2019-05-01T00:00:00.000'),
            endDate: new Date('2019-05-31T24:00:00.000'),
        },
        {
            id: '201906',
            name: 'June 2019',
            startDate: new Date('2019-06-01T00:00:00.000'),
            endDate: new Date('2019-06-30T24:00:00.000'),
        },
    ]
    const periodOnChange = {
        id: '201904',
        name: 'April 2019',
        startDate: new Date('2019-04-01T00:00:00.000'),
        endDate: new Date('2019-04-30T24:00:00.000'),
        level: 7,
        levelRank: 0,
        type: 'MONTHLY',
    }
    const onChangeSpy = jest.fn()
    let props

    beforeEach(() => {
        props = {
            periodId,
            period,
            periods,
            onChange: onChangeSpy,
            classes: {},
        }
    })

    it('should render timeline', () => {
        const { container } = render(<Timeline {...props} />)
        expect(container).toMatchSnapshot()
    })

    it('should call onChange with the period clicked', async () => {
        render(<Timeline {...props} />)

        // Find the first rect element
        const rects = screen.getAllByRole('button')
        const firstRect = rects[0]

        await fireEvent.click(firstRect)

        expect(onChangeSpy).toHaveBeenCalledWith(periodOnChange)
    })

    it('Should toggle play mode when play/pause button is clicked', async () => {
        render(<Timeline {...props} />)

        // Find the play/pause button
        const playPauseBtn = screen.getByRole('button', { name: /Play/i })

        // Initially, the play icon should be visible, and the pause icon should not
        expect(screen.queryByTestId('play-button')).toBeInTheDocument()
        expect(screen.queryByTestId('pause-button')).not.toBeInTheDocument()

        // Simulate clicking the play/pause button
        await fireEvent.click(playPauseBtn)

        // After clicking, the pause icon should be visible, and the play icon should not
        expect(screen.queryByTestId('play-button')).not.toBeInTheDocument()
        expect(screen.queryByTestId('pause-button')).toBeInTheDocument()

        // Ensure the onChangeSpy was called with the correct argument
        expect(onChangeSpy).toHaveBeenCalledWith(periodOnChange)

        // Simulate clicking the play/pause button again
        await fireEvent.click(playPauseBtn)

        // After clicking again, the play icon should be visible, and the pause icon should not
        expect(screen.queryByTestId('play-button')).toBeInTheDocument()
        expect(screen.queryByTestId('pause-button')).not.toBeInTheDocument()
    })
})
