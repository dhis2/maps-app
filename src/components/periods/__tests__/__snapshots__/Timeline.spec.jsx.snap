// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline should render timeline 1`] = `
<div>
  <svg
    class="dhis2-map-timeline timeline"
    style="height: 40px; bottom: 30px;"
  >
    <g
      aria-label="Play/Pause"
      class="play"
      role="button"
      transform="translate(7, 4)"
    >
      <path
        d="M0 0h24v24H0z"
        fill-opacity="0.0"
      />
      <path
        class="play-icon"
        d="M8 5v14l11-7z"
        data-test="play-button"
      />
    </g>
    <g
      transform="translate(40, 10)"
    >
      <rect
        class="period"
        height="8"
        role="button"
        width="-19.78021978021978"
        x="0"
        y="0"
      />
      <rect
        class="period"
        height="8"
        role="button"
        width="-20.439560439560438"
        x="-19.78021978021978"
        y="0"
      />
      <rect
        class="period selected"
        height="8"
        role="button"
        width="-19.78021978021978"
        x="-40.21978021978022"
        y="0"
      />
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(40, 20)"
    >
      <path
        class="domain"
        d="M0.5,6V0.5H-59.5V6"
        stroke="currentColor"
      />
    </g>
  </svg>
</div>
`;
